src/aggregates/vector_selector.rs

```diff
@@ -0,0 +1,404 @@
+) -> Option<Inner<VectorSelector>> {
+    unsafe {
+        in_aggregate_context(fcinfo, || {
+            match (state1, state2) {
+                (None, None) => None,
+                (None, Some(state2)) => {
+                    let s = state2.clone();
+                    Some(s.into())
+                }
+                (Some(state1), None) => {
+                    let s = state1.clone();
+                    Some(s.into())
+                } //should I make these return themselves?
+                (Some(state1), Some(state2)) => {
+                    let mut s1 = state1.clone(); // is there a way to avoid if it doesn't need it
+                    s1.combine(&state2);
+                    Some(s1.into())
+                }
+            }
+        })
+    }
+}
+
+extension_sql!(
+    r#"
+CREATE AGGREGATE @extschema@.vector_selector(
+    start_time TIMESTAMPTZ,
+    end_time TIMESTAMPTZ,
+    bucket_width BIGINT,
+    lookback BIGINT,
+    sample_time TIMESTAMPTZ,
+    sample_value DOUBLE PRECISION)
+(
+    sfunc = vector_selector_transition,
+    stype = internal,
+    finalfunc = vector_selector_final,
+    combinefunc = vector_selector_combine,
+    serialfunc = vector_selector_serialize,
+    deserialfunc = vector_selector_deserialize,
+    parallel = safe
+);
+"#,
+    name = "create_vector_selector_aggregate"
+);
+
+// The internal state consists of a vector non-overlapping sample buckets. Each bucket
+// has a corresponding (virtual) timestamp corresponding to the ts series
+// described above. The timestamp represents the maximum value stored in the
+// bucket (inclusive). The minimum value is defined by the maximum value of
+// the previous bucket (exclusive).
+// the value stored inside the bucket is the last sample in the bucket (sample with highest timestamp)
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct VectorSelector {
+    first_bucket_max_time: pg_sys::TimestampTz,
+    last_bucket_max_time: pg_sys::TimestampTz,
+    end_time: pg_sys::TimestampTz, //only used for error checking
+    bucket_width: Milliseconds,
+    lookback: Milliseconds,
+    elements: Vec<Option<(pg_sys::TimestampTz, f64)>>,
+}
+
+impl VectorSelector {
+    pub fn new(
+        start_time: pg_sys::TimestampTz,
+        end_time: pg_sys::TimestampTz,
+        bucket_width: Milliseconds,
+        lookback: Milliseconds,
+    ) -> Self {
+        let num_buckets = ((end_time - start_time) / (bucket_width * USECS_PER_MS)) + 1;
+
+        let last_bucket_max_time =
+            end_time - ((end_time - start_time) % (bucket_width * USECS_PER_MS));
+
+        VectorSelector {
+            first_bucket_max_time: start_time,
+            last_bucket_max_time: last_bucket_max_time,
+            end_time: end_time,
+            bucket_width: bucket_width,
+            lookback: lookback,
+            elements: vec![None; num_buckets as usize],
+        }
+    }
+
+    fn combine(&mut self, other: &Inner<VectorSelector>) {
```

Comment:
is this function ever hit in the tests? i.e. are we ever using a parallel plan in tests? can we force parallel plans?

Reply:
It isn't, thanks for catching that! I've added a test which forces execution of a parallel query, which should hit this function.